wnprc_billing/src/org/labkey/wnprc_billing/notification/WNPRCBillingNotificationProvider.java

```diff
@@ -133,4 +142,158 @@ public List<StringBuilder> getAdditionalNotifications(User user)
+
+        // iterate through Procedure queries associated with a chargeId
+        for (Map<String, Object> row : procedureQueries)
+        {
+            String procedureSchema = (String) row.get("schemaName");
+            String procedureQuery = (String) row.get("queryName");
+            String description = (String) row.get("description");
+            int chargeId = (Integer) row.get("chargeId");
+
+            // for each procedure query, get query results
+            UserSchema schema = QueryService.get().getUserSchema(u, c, procedureSchema);
+            TableInfo procedureQueryTi = schema.getTable(procedureQuery, null);
+            if (null == procedureQueryTi)
+            {
+                continue;
+            }
+
+            //filter by startDate and endDate
+            SimpleFilter filter = new SimpleFilter();
+            filter.addCondition(FieldKey.fromString("date"), startDate, CompareType.DATE_GTE);
+            filter.addCondition(FieldKey.fromString("date"), endDate, CompareType.DATE_LTE);
+            TableSelector procedureQueryTs = new TableSelector(procedureQueryTi, filter, null);
+            Collection<Map<String, Object>> procedureRows = procedureQueryTs.getMapCollection();
+
+            Double totalCostPerCategory = 0.00;
+
+            //iterate through procedure query results and get total cost
+            for (Map<String, Object> procedureRow : procedureRows)
+            {
+                // for each row, associate chargeId
+                procedureRow.put("chargeId", chargeId);
+
+                // add charge info
+                ChargeInfo ci = getChargeInfo(chargeId, chargeInfoArrayList, procedureRow.get("date"));
+
+                // get cost
+                Double unitCost = ci.getUnitCost();
+
+                // total cost
+                Double totalCost = unitCost * (Double) procedureRow.get("quantity");
+
+                totalCostPerCategory += totalCost;
+            }
+            Map<String, Double> totalsMap = new HashMap<>();
+            totalsMap.put("total", (procedureRows.size()*1.00));
+            totalsMap.put("totalCost", totalCostPerCategory);
+            totalsByCategory.put(description, totalsMap);
+            ActionURL url = QueryService.get().urlFor(u, c, QueryAction.executeQuery, procedureSchema, procedureQuery);
+
+            SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd");
+            url.addParameter("query.date~dategte", format.format(startDate));
+            url.addParameter("query.date~datelte", format.format(endDate));
+            additionalCategoryUrls.put(description, url.toString());
+        }
+    }
+
+    private ChargeInfo getChargeInfo(int chargeId, ArrayList<ChargeInfo> chargeInfoArrayList, Object date)
+    {
+        Date chargeDate = (Date) date;
+        List<ChargeInfo> chargeInfoList = chargeInfoArrayList.stream().filter(ci -> ci.getChargeId() == chargeId).toList();
+
+        if (chargeInfoList.size() == 1)
```

Comment:
I don't think we need either of these `if` statements. The `for` loop on its own will behave identically.

Reply:
yup, this was copilot code, which technically is correct so didn't put much thought into collapsing it further.